aip/general/0180.md

```diff
@@ -139,6 +139,84 @@ code, as such changes will be seen as breaking by those users.
 whether a proposed change is likely to break users, and an expansive reading of
 this guidance could ostensibly prevent _any_ change (which is not the intent).
 
+#### Default values must not change
+
+Changing the default value is considered breaking and **must not** be done. The
+default behavior for a resource is determined by it's default values, and this
+**must not** change across minor versions.
```

Comment:
Wanted to sanity check with @loudej on this one.

The motivation and client-friendliness makes total sense to me. At the same time, I'm wondering about how we do enable use cases like modifying the default disk type / instance type for compute APIs. New instance types come out all the time, and using them as the default helps GCP customers get more performance / $. 

One way I could imagine that being solved is by having clients always intern the defaults, and having servers *never* set the default. WDYT? 

Maybe if that was the solution then this guidance would still hold (don't change the server-side default).

Reply:
Are direct API clients relying heavily on defaults? Changes [outside of controlled windows] are probably more harm than good for IaC. I would consider having gcloud/Console change defaults clientside at will, and updating IaC during breaking change windows (or never).

If volatility across all clients is desirable, a double simplex (whether client or server side, although server side would be a little nicer imo. with obvious bias.) would be viable.

Reply:
> One way I could imagine that being solved is by having clients always intern the defaults, and having servers never set the default. WDYT?

So essentially mark all fields with defaults as required instead and bake the defaults into clients? It could work, but may have other usability issues like requiring a much larger set of values to be specified if a user is working with an API outside of a client.

Overall I think the suggestion of a double simplex can improve this situation greatly, as it would remove the direct connection between the unset value in an IaC client and the server-set default. Without that connection I expect we wouldn't see the issues that cause data loss when these defaults change.

Reply:
There are multiple points here.

1. Default values can't really change without a new major version unless the default value was always specified to be a dynamically chosen default, in which case the value must be pinned for a specific resource. It shouldn't change once the resource has been created. Double simplex is probably the best option for such cases, but this approach shouldn't be applied to every field.

2. The values need to be easily discoverable, such as for policy enforcement. In Kubernetes we didn't want admission controllers, asynchronous controllers, and other clients to need to try to figure out what default values were or how they were set or whether they were even defaults. Thus, Kubernetes explicitly sets all values that need to have values. However, that requires clients to gracefully accommodate that behavior. Server-side apply was necessary for Terraform integration. One idea we've discussed is (optionally) returning these values in validate_only responses and with a special Get request parameter.

3. Always setting the defaults in clients seems like a recipe for inconsistency and maintenance problems to me. Plus it doesn't work for attributes that are newer than the clients. Effectively it creates a "blueprint", regardless of what type of client/surface was used to create the resource and whether it was hardcoded or data-driven.

We need to provide more complete guidance around unset values, default values, dynamically generated / allocated / assigned values, and other values transformed by the service (semantic strings, associative lists, etc.), but I'm comfortable with the statement that changing default values or explicitly setting default values when they weren't previously populated explicitly are non-backward-compatible changes within the same major version.

Reply:
I'll raise my hand and say I don't know what's meant by "double simplex" here.

Just to establish baselines, can we all agree that @bgrant0607's statement of:

> It shouldn't change once the resource has been created.

... is correct? If a resource has been created, and the server has populated the field, it mustn't change to some other value later. I suspect we're all happy with that.

My guess is that the other side of Brian's point 1 (the viability of "if the client doesn't specify a value for this field, the server may choose a default, and that default may vary over time) will be less unanimously agreed. Would it be okay if the server guaranteed to always use the same value *if the client specifies the same API version*? (Looking to future work.)

I don't quite follow the point around policy enforcement, but I do agree it would be nice to be able to access the defaults... although that could be tricky if "default for field X depends on field Y".

Putting everything client-side does feel problematic - if we can agree on that much, we don't need to investigate the details of that further.

Reply:
> Default values can't really change without a new major version unless the default value was always specified to be a dynamically chosen default, in which case the value must be pinned for a specific resource. It shouldn't change once the resource has been created. 

Yes, this make sense to me and agreed. Although re-creating a resource may change the default value. The difference comes down to whether we do this in a major version of the API or not.

> Server-side apply was necessary for Terraform integration. One idea we've discussed is (optionally) returning these values in validate_only responses and with a special Get request parameter.

This is an interesting idea, although it does result in a apply for a Terraform resource being inconsistent if the server-side default changes. 

My motivation for integrating this into the client side was that then the client will always set the same default, and the behavior doesn't become dependent on the server. 

I do agree it can make the default case complicated for consumers. But even with a major version change, a customer would still have to deal with different clients having different defaults (unless they're all using the same major version of the API).

> Always setting the defaults in clients seems like a recipe for inconsistency and maintenance problems to me. Plus it doesn't work for attributes that are newer than the clients. Effectively it creates a "blueprint", regardless of what type of client/surface was used to create the resource and whether it was hardcoded or data-driven.

I think that's unavoidable no matter how you deal with a changing default. A major version change that only changes the default has the same effect. Although I do agree that a major version revision would be better in informing the user of a change vs a minor version.

> ... is correct? If a resource has been created, and the server has populated the field, it mustn't change to some other value later. I suspect we're all happy with that.

As a client I'm happy with that. As a service producer I worry that there will be a business need to change the default value, and we won't be able to enforce this guidance. 

> but I'm comfortable with the statement that changing default values or explicitly setting default values when they weren't previously populated explicitly are non-backward-compatible changes within the same major version.

The latter i'm comfortable with. The former has happened (compute instance disk type), and I think for compute this will happen regularly (default instance type changes).

Reply:
to summarize my stance: I don't agree with the principle, I just worry it's not defensible because this happens often enough to be a practice. e.g. default instance type for VM creation for compute.

That said it looks like GCE [hard-codes disk types in the client](https://cloud.google.com/compute/docs/disks#disk-types). And [instance types don't have defaults at all](https://cloud.google.com/compute/docs/reference/rest/v1/instances/insert). I guess I was thinking of AWS. 

I'm still fairly skeptical of enforcement but I agree with the rationale here.  We can try enforcing this guidance for now and see how it goes.

Reply:
GCE doesn't follow the AIPs. They explicitly set defaults in the resources, more like Kubernetes.

Something like the machine type or disk interface is actually in the category of "dynamic defaults". It's dynamically selected, allocated, computed/generated, etc. The "default" may change slowly in practice, but it could be different across regions or set via project-level settings or something as well as changing over time. 

"double simplex" was used in Greg's exploratory doc to describe the pattern of 2 fields, one input and one output. The output field would always return the effective value used by the system. The input value would respect the rules about not changing what was provided by the client.

Reply:
@jskeet re:
> Would it be okay if the server guaranteed to always use the same value if the client specifies the same API version?

Likely yes, but it may depend on the exact details of the API versioning scheme. 

The double simplex idea is an interesting way to mitigate these problems going forward, but I don't think it has much bearing on this proposal which is targeted at preventing breaking changes in existing APIs. We do need to come up with a plan for better handling of default values overall for new APIs.

For existing APIs though we have limited options and I believe the best path is to stop changes in server-side default values for existing APIs. Clients can modify client-set default values as appropriate to accommodate changing preferences and more efficient instance types in a non-breaking way. APIs can also change defaults during a major version increment.

Reply:
Having reread all of this, I'm pretty sure that:

- We need a way of defaults changing over time (think encryption algorithms etc)
- Some clients (e.g. Terraform) need a level of stability

Personally I'd be okay with having this guidance for now, with an understanding that when client-specific versioning happens, we change the guidance to "changing the default requires a new version, and the server will still honor the previous default when the client specifies the old version". Terraform can then consider "we've changed default" to require a new major Terraform provider version.

I think that's the best compromise we'll be able to reach, but I'd like to check that both @slevenick and @toumorokoshi are okay with it.